tests/unit/general/test_umpire_mem.cpp

```diff
@@ -77,7 +79,10 @@ static void test_umpire_device_memory()
    Vector host_perm(num_elems);
    REQUIRE(!is_pinned_host(host_perm.GetData()));
    // allocate on host, use temporary device memory when needed
-   Vector host_temp(num_elems); host_temp = host_val; host_temp.UseTemporary(true);
+   // (TODO: make sure this does not do the device allocation, i.e. use lazy
+   //        device allocations)
+   Vector host_temp(num_elems, MemoryType::DEVICE_UMPIRE_2);
+   host_temp = host_val; // done on host since UseDevice() is not set
```

Comment:
+1 for lazy device allocations.
We may also want to switch the device type after initialization, which was what this example was demonstrating (although that wasn't very clear, I should have documented it), do you have some kind of `UpdateDeviceType` in mind? Alternatively we can add more methods for setting the `MemoryType` in the mfem classes we are using but an update method seems less invasive

Reply:
Do you need to use the same vector with two different device memory types? E.g. first you use it with "non-pool" device memory and later you want to move it to "pool" device memory?

Reply:
> first you use it with "non-pool" device memory and later you want to move it to "pool" device memory?

We don't have that use case anywhere.

It may be enough to have a constructor with `h_mt` and `d_mt`, your `MFEM_WITH_DEVICE_MEMORY` idea, and lazy device allocations, I'll look more closely next week